### PR TITLE
Test | Fix Unit Tests for GetSqlServerSPN 

### PR DESCRIPTION
**Pull request by arellegue**
This PR is for the addition of Unit Tests of the GetSqlServerSPNs function of SNIProxy class.

The unit tests is to ensure that the GetSqlServerSPNs function returns a port number for TCP protocol and returns an instance name for NP protocol.

In addition, unit tests were added to test the DataTestUtility.ParseDataSource function.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs [hunk 1]
```diff
@@ -98,11 +98,24 @@ public static void PortNumberInSPNTestForTCP()
...
             builder.DataSource = $"{builder.DataSource},{port}";
 
-            PortNumberInSPNTest(builder.ConnectionString, port);
+            PortNumberInSPNTest(connectionString: builder.ConnectionString, expectedPortNumber: port);
+        }
+
+        [ConditionalFact(nameof(IsSPNPortNumberTestForNP))]
+        public static void SPNTestForNPMustReturnNamedInstance()
+        {
+            string connectionString = DataTestUtility.NPConnectionString;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:107]: If you're not really making a connection to target database, this should be a Functional (Unit) test where you provide deterministic connection string properties and compare with expected results from the function called.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs [hunk 2]
```diff
@@ -800,5 +809,17 @@ private bool InferNamedPipesInformation()
 
         private static bool IsLocalHost(string serverName)
             => ".".Equals(serverName) || "(local)".Equals(serverName) || "localhost".Equals(serverName);
+
+        private string GetInstanceNameFromDataSource()
+        {
+            string instanceName = "";
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs:815]: Applied the suggested usage of string.Empty.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs [hunk 3]
```diff
@@ -98,11 +98,24 @@ public static void PortNumberInSPNTestForTCP()
...
+            SqlConnectionStringBuilder builder = new(connectionString);
+
+            string instanceName = "";
+            DataTestUtility.ParseDataSource(builder.DataSource, out _, out _, out instanceName);
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:111]: Applied the suggested declaration of instanceName within the function call parameters.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs (hunk 1 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:107]: The unit test depends on the config.json that is only available in Manual test. I'm not sure how to get the DataTest.Utility within the Functional Test.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:107]: That's the point, it should not depend on Config.json if you're not actually connecting to anything. You can supply any dummy connection string in functional tests like we do in unit testing.

Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:107]: I'm sorry if I am misunderstanding this unit test.  However, what I know is that the unit test needs to get the result of the driver's GetSqlServerSPN function from an actual physical named instance sql server. Hence, it needs the named instance data source from config.json.  I will find out if it is possible to hard code the actual physical named instance sql server data source in a dummy connection string then I will change it to functional test.

Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:107]: @cheenamalhotra While the test doesn't connect to the database, it does use DNS to look up the server's FQDN in order to resolve the SPN for integrated security. It needs a real server for that.

Perhaps we should add code comments to make that clear.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:107]: That makes sense, I didn't see the `GetSPNInfo()` before.. was mainly looking at testing instance name from the PR context. No worries, thanks for clarifying.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs [hunk 4]
```diff
@@ -652,6 +652,11 @@ private bool InferConnectionDetails()
                 }
 
                 Port = port;
+
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs:655]: Please include a comment..
`// If user is connecting to a managed instance with port, e.g. localhost\INSTANCENAME,1434`

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs [hunk 5]
```diff
@@ -800,5 +809,17 @@ private bool InferNamedPipesInformation()
...
         private static bool IsLocalHost(string serverName)
             => ".".Equals(serverName) || "(local)".Equals(serverName) || "localhost".Equals(serverName);
+
+        private string GetInstanceNameFromDataSource()
+        {
+            string instanceName = string.Empty;
+            string[] tokensByBackSlash = _dataSourceAfterTrimmingProtocol.Split(BackSlashCharacter);
+            if (tokensByBackSlash.Length > 1)
+            {
+                instanceName = tokensByBackSlash[1];
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs:819]: Use Trim()

Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs:819]: There are also other places in the class, where tokens are not trimmed, can you fix those too?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs [hunk 6]
```diff
@@ -98,11 +98,23 @@ public static void PortNumberInSPNTestForTCP()
...
+            PortNumberInSPNTest(connectionString: builder.ConnectionString, expectedPortNumber: port);
+        }
+
+        [ConditionalFact(nameof(IsSPNPortNumberTestForNP))]
+        public static void SPNTestForNPMustReturnNamedInstance()
+        {
+            string connectionString = DataTestUtility.NPConnectionString;
+            SqlConnectionStringBuilder builder = new(connectionString);
+
+            DataTestUtility.ParseDataSource(builder.DataSource, out _, out _, out string instanceName);
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:110]: Please unit test this particular method with wide variety of datasource names.
Include multiple test cases to validate parser with/without instancename/port/protocol etc.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs (hunk 4 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs:655]: Suggested comment was added.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs (hunk 5 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs:819]: Suggested usage of Trim() function was added.

Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs:819]: Found all usage of token then use Trim() function when used.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs (hunk 6 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/InstanceNameTest/InstanceNameTest.cs:110]: Unit tests for DataTestUtility.ParseDataSource were added for a total of 24 tests. Let me know if you want to add more tests.

**Review by DavoudEshtehari — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataSourceParserTest/DataSourceParserTest.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            Assert.Equal("localhost", hostname);
+        }
+
+        [ConditionalTheory(typeof(DataTestUtility), nameof(DataTestUtility.IsNotAzureServer), nameof(DataTestUtility.IsNotAzureSynapse))]
+        [InlineData("localhost,1433")]
+        [InlineData("tcp:localhost,1433")]
+        [InlineData("np:localhost,1433")]
+        [InlineData(" localhost,1433 ")]
+        [InlineData(" tcp:localhost,1433 ")]
+        [InlineData(" np:localhost,1433 ")]
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataSourceParserTest/DataSourceParserTest.cs:36]: NIT: these are for trim test? If yes, one of them should suffice

Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataSourceParserTest/DataSourceParserTest.cs]: For completeness sake, both specified protocols, tcp and np, were included in all the DataSourceParser tests.

As you pointed out, the parsing of both tcp and np protocols will both result in the same behavior for parsing data source with space and colon in it.  Thus, one of them should suffice. The np protocol data source has been dropped in all unit tests of DataSourceParser.

**Review by JRahnama — APPROVED** [review]

**Pull request merged**